src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -2670,5 +2671,260 @@ private interface INegator<T> where T : struct
             public static Vector128<T> NegateIfNeeded(Vector128<T> equals) => ~equals;
             public static Vector256<T> NegateIfNeeded(Vector256<T> equals) => ~equals;
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void Reverse<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 0);
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                if (Unsafe.SizeOf<T>() == sizeof(byte))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, byte>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(char))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, short>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(int))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, int>(ref elements), length);
+                    return;
+                }
+                else if (Unsafe.SizeOf<T>() == sizeof(long))
+                {
+                    ReverseSIMD(ref Unsafe.As<T, long>(ref elements), length);
+                    return;
+                }
+            }
+            ReverseScalar(ref elements, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseScalar<T>(ref T elements, nuint length)
+        {
+            if (length <= 1)
+                return;
+            ref T first = ref elements;
+            ref T last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, (int)length), 1);
```

Comment:
This type of logic is incredibly hard to read/understand and has led to bugs in the past.

It is likely better to just pin and use pointers here instead. CC @jkotas 

Notably in this particular case, it would be much more readable as just `ref Unsafe.Add(ref first, length - 1)` (especially since we have an overload that takes `nuint`)

Reply:
> `ref Unsafe.Add(ref first, length - 1)`

I agree that this is more readable.

> It is likely better to just pin and use pointers here instead. CC @jkotas

I am ok with using byref arithmetic in small methods like this one where it is easy to review the code for correctness.